presto-main/src/main/java/com/facebook/presto/server/AsyncPageTransportServlet.java

```diff
@@ -0,0 +1,196 @@
+    }
+
+    @Override
+    protected void doGet(HttpServletRequest request, HttpServletResponse response)
+            throws IOException
+    {
+        String requestURI = request.getRequestURI();
+        // Example:  /v1/task/async/{taskId}/results/{bufferId}/{token}
+        List<String> requestURIParts = Arrays.asList(requestURI.split("/"));
+
+        if (requestURIParts.size() != 8) {
+            response.sendError(SC_BAD_REQUEST, format("Unexpected URI for task result request in async mode: %s", requestURI));
+            return;
+        }
+
+        TaskId taskId = TaskId.valueOf(requestURIParts.get(4));
+        OutputBufferId bufferId = OutputBufferId.fromString(requestURIParts.get(6));
+        long token = parseLong(requestURIParts.get(7));
+        DataSize maxSize = DataSize.valueOf(request.getHeader(PRESTO_MAX_SIZE));
+
+        AsyncContext asyncContext = request.startAsync(request, response);
+        // wait time to get results
+        Duration waitTime = randomizeWaitTime(DEFAULT_MAX_WAIT_TIME);
+        asyncContext.setTimeout(waitTime.toMillis() + pageTransportTimeout.toMillis());
+
+        asyncContext.addListener(new AsyncListener()
+        {
+            public void onComplete(AsyncEvent event)
+            {
+            }
+
+            public void onError(AsyncEvent event)
+                    throws IOException
+            {
+                String errorMessage = format("Server error to process task result request %s : %s", requestURI, event.getThrowable().getMessage());
+                log.error(event.getThrowable(), errorMessage);
+                response.sendError(SC_INTERNAL_SERVER_ERROR, errorMessage);
+            }
+
+            public void onStartAsync(AsyncEvent event)
+            {
+            }
+
+            public void onTimeout(AsyncEvent event)
+                    throws IOException
+            {
+                String errorMessage = format("Server timeout to process task result request: %s", requestURI);
+                log.error(event.getThrowable(), errorMessage);
+                response.sendError(SC_INTERNAL_SERVER_ERROR, errorMessage);
+            }
+        });
+
+        ListenableFuture<BufferResult> bufferResultFuture = taskManager.getTaskResults(taskId, bufferId, token, maxSize);
+        bufferResultFuture = addTimeout(
+                bufferResultFuture,
+                () -> BufferResult.emptyResults(taskManager.getTaskInstanceId(taskId), token, false),
+                waitTime,
+                timeoutExecutor);
+
+        ServletOutputStream out = response.getOutputStream();
+        addCallback(bufferResultFuture, new FutureCallback<BufferResult>()
+                {
+                    @Override
+                    public void onSuccess(BufferResult bufferResult)
+                    {
+                        Queue<SerializedPage> serializedPages = new LinkedList<>(bufferResult.getSerializedPages());
```

Comment:
I think we could encapsulate the queue creation inside the `SerializedPageWriteListener`.  Also, why not use an `ArrayDequeue`?

Reply:
`Deque` extends `Queue`, and the writer interface actually doesn't need `Deque`, so i think just as a mater of interface lets stick to a "tighter" one. As for the implementation - `ArrayDeque` should be more efficient for this use case, but given that the pages are expected to be large - it shouldn't make a lot of difference.

Reply:
My suggestion wasn't to change to the `Deque` interface--we can keep the variables type.  I don't want to block the PR on this suggestion, it's a minor nit as I did understand it won't make a large performance difference, it was more a suggestion that `ArrayDequeue` _in general_ is a better default when one wants to reach for a `Queue` as it performs better.

Reply:
That makes sense. I think we can change the implementation to the ArrayDeque